src/request.rs

```diff
@@ -0,0 +1,178 @@
+use std::io::{self, Read};
+use std::time::{Duration, Instant};
+use error::Error;
+use super::Request;
+
+use httparse;
+
+fn elapsed_milliseconds(from: &Instant) -> u64 {
+    let elapsed = Instant::now() - *from;
+    (elapsed.as_secs() * 1000) + (elapsed.subsec_nanos() as u64 / 1_000_000)
+}
+
+fn duration_to_milliseconds(from: &Duration) -> u64 {
+    (from.as_secs() * 1000) + (from.subsec_nanos() as u64 / 1_000_000)
+}
+
+pub(super) fn read<'request, S: Read>(
+    buf: &'request mut [u8],
+    stream: &mut S,
+    timeout: Option<Duration>,
+) -> Result<Request<&'request [u8]>, Error> {
+    let start_time = Instant::now();
+    let mut total_read = 0;
+
+    loop {
+        if total_read == buf.len() {
+            return Err(Error::RequestTooLarge);
+        }
+
+        let read = match stream.read(&mut buf[total_read..]) {
+            Ok(num) if num == 0 => return Err(Error::ConnectionClosed),
```

Comment:
is the connection closed really an error? that's why you `doc(hidden)`'d it, right?

Reply:
I actually marked `RequestIncomplete` as `doc(hidden)`. There was no specific reason other than the error shouldn't bubble up. It will just result in a `413 Payload Too Large` response. By that logic though, I should have marked `ConnectionClosed` as hidden too. D'oh!